lib/kennel/utils.rb

```diff
@@ -105,18 +105,22 @@ def parallel(items, max: 10)
         threads = [items.size, max].min
         work = items.each_with_index.to_a
         done = Array.new(items.size)
+        mutex = Mutex.new
         workers = Array.new(threads).map do
           Thread.new do
             loop do
-              item, i = work.pop
+              item, i = mutex.synchronize { work.pop }
               break unless i
-              done[i] =
+              result =
                 begin
                   yield item
                 rescue StandardError => e
-                  work.clear
                   e
                 end
+              mutex.synchronize do
+                done[i] = result
+                work.clear if e.is_a?(StandardError)
+              end
```

Comment:
it does not matter when we clear / there is no race condition afaik

Reply:
And yet kennel gets stuck when running the github actions – only the ones involving this threading.

Reply:
```
workers.each(&:join)
lib/kennel/utils.rb:123:in `join'
```
from 
```
parts = Utils.parallel(projects, &:validated_parts).flatten(1)
lib/kennel.rb:121
```

... and in the thread it's

```
gems/zeitwerk-2.6.0/lib/zeitwerk/kernel.rb:27:in `require'
```

so something about auto-load
the other sus thing was that we rescue StandardError and not Exception but did not set abort_on_exception so good chance there is something going on there too
fix for that https://github.com/grosser/kennel/pull/177

still it seems like the slow tests hang in random places, so maybe there is something eating all the cpu and that is why it's slow